### PR TITLE
add more cli tests for kubens

### PR DESCRIPTION
**Pull request by doodlesbykumbi**

### CODE REVIEW

**Comment by ahmetb** [comment]
I think you pushed to the branch of #116 so I can't merge that now.

File: kubens [hunk 1]
```diff
@@ -118,6 +118,8 @@ choose_namespace_interactive() {
 set_namespace() {
   local ctx prev
   ctx="$(current_context)" || exit_err "error getting current context"
+  [[ -z "$ctx" ]] && exit_err "current context is empty"
```
Comment (ahmetb) [inline kubens]: I can't imagine a case where this returns empty string.  Can you clarify? Ideally the previous line should fail.

File: kubens [hunk 2]
```diff
@@ -173,14 +175,19 @@ swap_namespace() {
 }
 
 main() {
-  if hash kubectl 2>/dev/null; then
+  set +u
```
Comment (ahmetb) [inline kubens]: don't do this. the many `set -u`s you have below clearly shows it'll cause trouble someday when we forget to add `set -u` to a new code path. :)

instead use a one-off default like:

    "${KUBECTL:-}"

in the IF statement.

**Comment by doodlesbykumbi** [comment]
@ahmetb i closed #116 and deleted the branch then created a brand new branch with all the changes. you should be good to merge from here.

File: kubens [hunk 3]
```diff
@@ -173,14 +175,19 @@ swap_namespace() {
...
+  if [[ ! -z "$KUBECTL" ]]; then
```
Comment (ahmetb) [inline kubens]: I believe this is equivalent to `-n`. :)

File: mock-kubectl [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
```
Comment (ahmetb) [inline mock-kubectl]: move this to `test/`

File: mock-kubectl [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+[[ -n $DEBUG ]] && set -x
+
+set -eou pipefail
+
+if hash kubectl 2>/dev/null; then
```
Comment (ahmetb) [inline mock-kubectl]: I don't understand why this code block is here. 

I think we should override KUBECTL only for calls that fail without a cluster (like `kubectl get namespace`); not the entire test suite.

that would simplify both the mock-kubectl script, and makes it easier to reason about the tests.

File: mock-kubectl [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  KUBECTL=kubectl
+elif hash kubectl.exe  2>/dev/null; then
+  KUBECTL=kubectl.exe
+else
+  set -u
+  echo >&2 "kubectl is not installed"
+  exit 1
+fi
+
+if [[ $@ == *' namespaces '* ]]; then
```
Comment (ahmetb) [inline mock-kubectl]: we can probably be more specific like `$@ == "get namespaces"`

File: test/common.bash [hunk 7]
```diff
@@ -17,6 +17,29 @@ use_config() {
...
+}
+
 get_context() {
-    kubectl config current-context
+  kubectl config view -o=jsonpath='{.current-context}'
+}
+
+exit_err() {
+  echo >&2 "${1}"
+  exit 1
```
Comment (ahmetb) [inline test/common.bash]: I'm not sure if `exit 1` is the way to fail from a bats test.

Can you just try making the last command `false` here, and see if it fails properly? I suspect if all `set -e`s are correct, this failure should propagate.

File: test/kubens.bats [hunk 8]
```diff
@@ -13,3 +16,90 @@ COMMAND="${BATS_TEST_DIRNAME}/../kubens"
...
+}
+
+@test "switch to previous namespace when none exists" {
+  use_config config1
+  switch_context user1@cluster1
+
+  run ${COMMAND} -
+  echo "$output"
+  [ "$status" -eq 1 ]
+  [[ "$output" = "error: No previous namespace found for current context." ]]
```
Comment (ahmetb) [inline test/kubens.bats]: I wouldn't compare error messages, these are prone to change a bit frequently.

Comment (doodlesbykumbi) [reply test/kubens.bats]: that may be true but i think it's still worth asserting on the output as much as on the exit status code. 
the test case seems empty to me without asserting on the output

File: test/common.bash (hunk 7 above)
Comment (doodlesbykumbi) [reply test/common.bash]: noted but turns out this wasn't necessary at all

File: mock-kubectl (hunk 6 above)
Comment (doodlesbykumbi) [reply mock-kubectl]: updated

File: mock-kubectl (hunk 5 above)
Comment (doodlesbykumbi) [reply mock-kubectl]: removed all this junk. it was unnecessary after all.

File: mock-kubectl (hunk 4 above)
Comment (doodlesbykumbi) [reply mock-kubectl]: noted and done

File: kubens (hunk 2 above)
Comment (doodlesbykumbi) [reply kubens]: 🤦‍♂️... thanks

**Comment by doodlesbykumbi** [comment]
@ahmetb addressed your comments and replied where it made sense to do so

File: kubens (hunk 1 above)
Comment (doodlesbykumbi) [reply kubens]: this was also not necessary

File: test/kubens.bats [hunk 9]
```diff
@@ -1,6 +1,9 @@
 #!/usr/bin/env bats
 
 COMMAND="${BATS_TEST_DIRNAME}/../kubens"
+export KUBECTL="$BATS_TEST_DIRNAME/../test/mock-kubectl"
```
Comment (ahmetb) [inline test/kubens.bats]: If we override KUBECTL _only_ when we need the "get namespaces" functionality, would that pretty much be _all_ test cases?

File: test/kubens.bats [hunk 10]
```diff
@@ -13,3 +16,90 @@ COMMAND="${BATS_TEST_DIRNAME}/../kubens"
   echo "$output">&2
   [ "$status" -eq 0 ]
 }
+
+@test "list namespaces when no kubeconfig exists" {
+  run ${COMMAND}
+  echo "$output"
+  [ "$status" -eq "1" ]
```
Comment (ahmetb) [inline test/kubens.bats]: I think we should use `[[` for consistency, no reason not to. 

also you used `"1"` here vs `0` above, please don't quote the numeric value.

File: test/mock-kubectl [hunk 11]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+[[ -n $DEBUG ]] && set -x
+
+set -eou pipefail
+
+if [[ $@ == *'get namespaces'* ]]; then
+  echo "default"
```
Comment (ahmetb) [inline test/mock-kubectl]: let's call these ns1 and ns2. no need to stick to original names (hard to follow in test cases above), and no need to have 3.

File: kubens [hunk 12]
```diff
@@ -173,7 +173,9 @@ swap_namespace() {
 }
 
 main() {
-  if hash kubectl 2>/dev/null; then
+  if [[ "${KUBECTL:-}" ]]; then
```
Comment (ahmetb) [inline kubens]: I don't know what `[[ "$var" ]]` test does (even though manpage says it checks string isn't null), I recommend you stick to -n or -z here.

Not sure what the next line saying `true` does either.

I recommend if you rewrite this like

```
if [[ -z "${KUBECTL:-}" ]] then;
  if hash ..
  elif ..
  else ..
  fi

fi
```

File: test/kubens.bats [hunk 13]
```diff
@@ -13,3 +16,90 @@ COMMAND="${BATS_TEST_DIRNAME}/../kubens"
...
+
+  run ${COMMAND}
+  echo "$output"
+  [ "$status" -eq 0 ]
+  [[ "$output" = *"default"* ]]
+  [[ "$output" = *"kube-public"* ]]
+  [[ "$output" = *"kube-system"* ]]
+}
+
+@test "switch to existent namespace" {
```
Comment (ahmetb) [inline test/kubens.bats]: existing*

File: test/kubens.bats (hunk 9 above)
Comment (doodlesbykumbi) [reply test/kubens.bats]: it seems that way for now. `get namespaces` is the only call to `kubectl` that requires a cluster

File: test/docker/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,10 @@
+FROM ruby:2.3.8
```
Comment (ahmetb) [inline test/docker/Dockerfile]: I think you checked these in accidentally :)

Comment (ahmetb) [reply test/docker/Dockerfile]: Ah, no thanks, please don't include changes that weren't discussed. :) Unless you're on windows, it's pretty easy to install bats. So I don't think we should add scripts and dockerfiles almost as long as the tests to run the tests.

Comment (doodlesbykumbi) [reply test/docker/Dockerfile]: Fair enough. Pruned :)

**Comment by ahmetb** [comment]
Thanks these are looking good now! 🙏
